app_dart/lib/src/model/firestore/task.dart

```diff
@@ -0,0 +1,214 @@
+// Copyright 2019 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:googleapis/firestore/v1.dart' hide Status;
+
+import '../../request_handling/exceptions.dart';
+import '../../service/firestore.dart';
+import '../../service/logging.dart';
+import '../luci/push_message.dart';
+
+class Task extends Document {
+  /// Lookup [Task] from Firestore.
+  ///
+  /// `documentName` follows `/projects/{project}/databases/{<}database}/documents/{document_path}`
+  static Future<Task> fromFirestore({
+    required FirestoreService firestoreService,
+    required String documentName,
+  }) async {
+    final Document document = await firestoreService.getDocument(documentName);
+    return Task.fromDocument(taskDocument: document);
+  }
+
+  /// Create [Task] from a task Document.
+  static Task fromDocument({
+    required Document taskDocument,
+  }) {
+    return Task()
+      ..fields = taskDocument.fields!
+      ..name = taskDocument.name!;
+  }
+
+  /// The task was cancelled.
+  static const String statusCancelled = 'Cancelled';
+
+  /// The task is yet to be run.
+  static const String statusNew = 'New';
+
+  /// The task failed to run due to an unexpected issue.
+  static const String statusInfraFailure = 'Infra Failure';
+
+  /// The task is currently running.
+  static const String statusInProgress = 'In Progress';
+
+  /// The task was run successfully.
+  static const String statusSucceeded = 'Succeeded';
+
+  /// The task failed to run successfully.
+  static const String statusFailed = 'Failed';
+
+  /// The task was skipped or canceled while running.
+  ///
+  /// This status is only used by LUCI tasks.
+  static const String statusSkipped = 'Skipped';
+
+  /// The list of legal values for the [status] property.
+  static const List<String> legalStatusValues = <String>[
```

Comment:
Would it make sense to define an enum with the string names and the legal status? Something like:

```
enum TaskStatus {

  statusFailed ('Failed', true, true),
  statusInProgress('In progress', true, false),
  ...
  statusInfraFailure('Infra Failure', true, true);

  const({
    required this.name,
    required this.isLegalStatus,
   required this.isTerminalStatus,
  });

  final String name;
  final bool isLegalStatus;
  final bool isTerminalStatus;
}
```

Reply:
This sounds a good idea. I expect it will need a bit refactoring of multiple places, and prefer to skip it from this PR. Created https://github.com/flutter/flutter/issues/143465 to track.

Reply:
Yeah no problem.